### ISSUE TITLE
Add support for configuration discovery

### ISSUE THREAD

**Issue by tsandall** [open-policy-agent/opa#987]
OPA currently supports a configuration file that defines the remote endpoints to download bundles from and upload status and decision logs to. In the future, OPA may gain new management functionality and the configuration would likely need to be extended. At the same time, it can be a pain to manage configuration files in a distributed environment.

To simplify deployment of OPA and enable better centralized management, we should extend OPA to discover it's configuration from a remote endpoint.

# High level design

* Implement a "discovery" package that downloads a bundle from a remote endpoint. The bundle will be extracted and evaluated locally by OPA. The result of evaluation should produce an OPA configuration.

* If discovery is enabled, execute the discovery in the OPA runtime at bootup (before plugins are started.) Take the result of the discovery execution (i.e., an OPA configuration) and load the plugins.

* Discovery can be enabled via configuration. For example:

```
services:
  - name: acmecorp
    url: https://example.com/control-plane-api/v1
    credentials: {bearer: {token: secret}}

discovery:
  path: /foo/bar
```

* To start, discovery should be mutually exclusive with other kinds of configuration: if discovery is on, bundles/decision logs/status cannot be configured manually.

# Details

* Conceptually, discovery can be implemented as a function that takes a configuration object and returns a configuration object. Discovery can be executed inside the initPlugins function in runtime.go immediately before the plugin manager is started. In the future, discovery could be a long-running process that reconfigures plugins if their configuration changes.

**Comment by repenno** [comment]
We should probably leverage https://tools.ietf.org/html/rfc5785 for this feature. OPA tries the well-known URI for the domain it is located which would provide the configuration or the URI of where to get the configuration

**Comment by repenno** [comment]
There are quite a number of services and application that use such approach 

https://en.wikipedia.org/wiki/List_of_/.well-known/_services_offered_by_webservers

**Comment by tsandall** [comment]
For now we will only support a simple HTTP-based discovery protocol outlined above. We can add support for additional service protocols in the future as needed.

**Comment by repenno** [comment]
Are you saying you will implement RFC5785 or not? I am a bit confused. RFC5785 is very simple, a proposed standard, it is HTTP based, and quite close to what you proposed.

**Comment by tsandall** [comment]
OPA is going to make a `GET` request to a remote HTTP server at a path defined in the configuration (e.g., in the example above, it would be `https://example.com/control-plane-api/v1/foo/bar`. I've edited the example slightly, i.e., it was a boolean flag before, but the concept is the same).  The remote HTTP server should respond with a bundle file. OPA will extract the bundle in-memory and evaluate the policy contained inside to obtain the configuration it should use. 

This approach has a few benefits.

1. The remote HTTP server could be an existing bundle server implementation. No server changes would be needed.
2. The discovery bundle is evaluated locally by OPA -- this will reduce the load on the remote server.
3. The discovery bundle is cacheable -- this will reduce latency and reduce load on the remote server.

If someone were to configure OPA to point at a /.well-known path, I think this approach would be compliant. The RFC doesn't say anything about the media type of the response or how the response is handled (which is the bulk of the work required here.)

**Comment by repenno** [comment]
If I do not configure a URL, would a OPA server go to  /.well-known/<hostname>/config so that everything works out of the box without any configuration?

**Comment by tsandall** [comment]
I think for now we should require a path. I haven't seen this .well-known scheme used in practice. If there's some strong benefit that I'm missing here, please let me know.

**Comment by repenno** [comment]
The wikipedia article has lots of references of big companies using this scheme together with their standard references.

**Comment by repenno** [comment]
We can register the OPA well-known name. We can write a small RFC and explain how to use it.

**Comment by tsandall** [comment]
This is implemented in v0.10.2.
